Documentation/proposals/202411-zone-aware-sharding.md

```diff
@@ -0,0 +1,227 @@
+assert(num_shards % len(zones) == 0) # Warning: multi-scraping of instances
+
+shards_per_zone := max(1, floor(num_shards / len(zones)))
+prom_zone_idx := shard_index % shards_per_zone
+    
+if zone_label == zones[prom_zone_idx] {
+    
+    assignment_idx := floor(shard_index / shards_per_zone)
+
+    if hash(address) % shards_per_zone == assignment_idx {
+        do_scrape()
+    }
+}
+```
+
+By using modulo to calculate the `prom_zone_idx`, instances will be distributed
+to zones in the sense of `A,B,C,A,B,C` and so on. This allows a modification of
+the `num_shards` value without redistribution of shards or data.
+This was preferred over allowing the number of `zones` to change, as this is
+less likely to happen.
+
+### Edge cases
+
+We have introduced asserts in the above section to warn about edge cases that
+might lead to duplicate data or data loss.
+
+By the above algorithm. prometheus instances will be distributed in an
+alternating fashion by using the already existing shard index.
+This leads to the following edge cases:
+
+When `num_shards` is 10 and `len(zones)` is 3 as in `[A..C]`.
+`shards_per_zone` is 3. This yields the following distribution
+
+```
+0 1 2 | 3 4 5 | 6 7 8 | 9 | shard index
+A B C | A B C | A B C | A | zone
+0 0 0 | 1 1 1 | 2 2 2 | 0 | assignment index
+```
+
+In this case the 2nd assert will warn about double scraping of instances in
+zone A, as the same targets are being assigned to instance 0 and 9.
+
+When `num_shards` is 2 and `len(zones)` is 3 as in `[A..C]`.
+`shards_per_zone` is 1. This yields the following distribution
+
+```
+0 1   | shard index
+A B C | zone
+0 0 0 | assignment index
+```
+
+In this case the 1st assert will warn about zone C not being scraped.
+
+## Required configuration options
+
+Following the algorithm presented above, we suggest the following configuration
+options to be added to the [Prometheus custom resource definition](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#monitoring.coreos.com/v1.Prometheus):
+
+```yaml
+spec:
+  shardingStrategy:
+    mode: 'classic'             # can be 'classic' or 'topology'
+    label: '__address__'        # defaults to '__address__` for 'classic' and 
+                                # 'topology.kubernetes.io/zone' for 'topology'
+    topology: []                # expects list of zones for 'toplogy'
+    additionalRelabelConfig: [] # Optional: array of relabel configurations
+```
+
+The `additionalRelabelConfig` section is meant to allow the `label` to be
+generated if needed. This should allow enough flexibility to cover edgecases
+not anticiapted by this proposal.
+
+## Generated configuration
+
+The following configuration uses Helm templating syntax and the proposed
+configuration structure as proposed above. Functions from the [sprig](https://masterminds.github.io/sprig/)
+library are used, as they are also present in Helm.
```

Comment:
I'm not sure to understand: are you saying that the configuration generated by the operator will processed by a template engine before being loaded into Prometeus? or is it only provided as an example?

Reply:
No, I just wanted to use some commonly used „language“ to express what should happen here.

I also wanted to use this to test the algorithm but I have not yet set up a test environment to actually do that.

Reply:
added an explanation to the section